### ISSUE TITLE
how should removal of preservation catalog entry for a storage root happen?

### ISSUE THREAD

**Issue by jmartin-sul** [sul-dlss/preservation_catalog#1400]
due to some mis-communication (and an assumption on my part that this would get captured in the manual operations described in https://github.com/sul-dlss/preservation_catalog/blob/master/.github/ISSUE_TEMPLATE/storage-migration-checklist.md), this has fallen through the cracks until now.

**NOTE** this issue is just about deleting the preservation catalog database entry for a storage root from which content has been migrated to new storage.  it is _not_ about actually sunsetting the physical storage or deleting the files from the old file system:  for that discussion, see #1327.

since pres cat does not currently fully support having the same druid multiple times among the moabs it tracks (`CompleteMoab` objects/`complete_moabs` rows), and since we're not immediately deleting content from the storage root from which content is being migrated, we must keep pres cat from knowing about that storage root anymore.  that way, it doesn't e.g. try to run M2C, and re-discover an old copy of a moab that's been migrated to a new storage root.

i had been assuming that this would be a manual process:  the person doing the migration (running through the above linked checklist) would:
* run the migration rake task
* make sure everything had been moved successfully
* make sure nothing still referenced the old storage root
* delete the database entry for the old `MoabStorageRoot`, as well as the corresponding shared configs entry.  note that db seeding is entirely additive, so removing an entry from shared_configs changes will not automatically remove it from the database.

the other option would be to have the rake task which migrates moabs from an old storage root also delete that old storage root's entry from preservation catalog's database.  i am wary of this option because:
* i _think_ i'd want the rake task to do the move in one transaction instead of a `#find_each`
* i'd definitely want the rake task to do some other integrity checking to make sure that everything had been moved, and that things were present on the file system as expected.  if the rake task weren't made more transactional, i think it'd have to query to see if it had successfully moved everything, but that sounds prone to error and the need for continually re-polling.
  * if we used the facilities of the `postgresql_cursor` gem that was leveraged for reporting, we probably could replace the `#find_each` with something that was entirely transactional but still performant.

still, for this migration, i think it'd be less effort to have a human use a checklist to do those things manually than it would be to code that into the rake task.

but i think this merits some discussion with more team members.  i'm certainly open to arguments that run counter to my current preference.

**Comment by jmartin-sul** [comment]
I ended up testing a different option when playing the part of ops for our dry run in #1420/#1377.

I would suggest that we essentially do the inverse of what ops is planning to do for the final storage root cutover as described in #1376.  during the `Ops Runs with the Baton` section of the checklist, i would suggest that ops do the following:
* in the **old** storage root directory, `mv` the `sdr2objects` directory to `sdr2objects_old`
* `mkdir` a sibling `sdr2objects` directory in the **old** storage root directory.  this new `sdr2objects` directory is a dummy directory, and should remain empty.  it is just there to provide a `storage_trunk` for the `moab-versioning` gem to find in the expected location, but it will remain unused until the entire **old** storage root is removed.

This has the following benefits:
* we don't need to remove the pres cat entry for the old `MoabStorageRoot` during the migration weekend
* which keeps us from having to `null` out the `from_moab_stoarge_root` field on the freshly migrated moabs, which would be necessary to delete the `MoabStorageRoot`, since there's a non-nullable foreign key relationship between the `complete_moabs.from_moab_stoarge_root` column to the `moab_storage_roots` table.
* but pres cat still won't pick up the moved moabs in their old location, because the storage trunk it will be looking in for that storage root won't actually contain anything anymore.  without having to immediately delete the freshly migrated content from disk.

Later in the week, when we have more certainty that the migration was a success, we can do the following to delete the old storage root entry from the catalog:
* `null` out the `from_moab_storage_root` field on the moabs that were migrated to the new storage.
* run `#destroy` on the `MoabStorageRoot` object for the _**old**_ storage root (e.g. `MoabStorageRoot.find_by(name: 'old_storage_root').destroy`).

I've tested this in stage, and it seems to work as expected.  When running `Stanford::MoabStorageDirectory.find_moab_paths(old_storage_root.storage_location)`, no paths are returned.  `find_moab_paths` is how M2C walks the storage root to return all the paths that look like druid trees.  see https://github.com/sul-dlss/preservation_catalog/blob/0be05727ae0fcfaa0a9690629cb183036fb67cb5/app/models/moab_storage_root.rb#L33-L37

Also as expected, when I run `Stanford::MoabStorageDirectory.find_moab_paths(new_storage_root.storage_location)`, I see the migrated moabs, plus all new content that has been flowing in, since that target storage root is still the latest in the configuration (and so the one to receive all new content).

question for @julianmorley:  is the `mv` and dummy replacement of the `sdr2objects` directory in the old storage root something that ops can indeed add to their checklist of work?  if so, should i file a ticket or create a PR somewhere, or will you guys handle that?  this is something that needs to happen to keep the subsequent automatic weekly M2C run from spewing errors, so i just want to make sure it doesn't fall through the cracks (esp since that might get us rate limited by honeybadger, which would be bad).

if that all sounds good, i will also create a small PR to add that to the infra team's migration checklist (just one of the bullet points in the `Ops Runs with the Baton` section).
